view/yura0302/src/component/BarChart.tsx

```diff
@@ -0,0 +1,120 @@
+import { useEffect, useRef, useState } from 'react';
+import * as d3 from 'd3';
+import { data } from '../data/data';
+
+interface ChartData {
+  currency: string;
+  values: { date: Date; value: number }[];
+}
+
+const BarChart = () => {
+  const svgRef = useRef<SVGSVGElement>(null);
+  const [chartData, setChartData] = useState<ChartData[]>([]);
+  const [allData, setAllData] = useState<ChartData[]>([]);
+
+  useEffect(() => {
+    const parseData = d3.csvParse(data);
+    const formatData: ChartData[] = parseData.map((d) => ({
+      currency: d.통화,
+      values: Object.keys(d)
+        .filter((key) => key !== '통화')
+        .map((key) => ({ date: new Date(key), value: +d[key] })),
+    }));
+    setAllData(formatData);
+  }, []);
```

Comment:
요 data는 BarChart 컴포넌트에 종속된 data가 아니라서 함수를 따로 만들어서 거기서 format 해주고 useState에서 initialize 해주면 좋을 것 같아요

```suggestion
// 별도 파일에서 유틸함수 생성
const getFormattedData = () => {
    const parseData = d3.csvParse(data);
    const formatData: ChartData[] = parseData.map((d) => ({
      currency: d.통화,
      values: Object.keys(d)
        .filter((key) => key !== '통화')
        .map((key) => ({ date: new Date(key), value: +d[key] })),
    }));
}

// useState initialize에서 사용 
const [chartData, setChartData] = useState<ChartData[]>(getFormattedData());
```

Reply:
제 코드 리뷰는 아니지만 코드 작성까지,, 정성 멋있습니다!!

Reply:
csv data 파싱하고 순회까지 도는 로직이니 추후 파싱해야하는 데이터 크기가 커진다면 `useDeferredValue` 훅을 고려해봐도 좋을 것 같습니다~

https://react.dev/reference/react/useDeferredValue